olp-cpp-sdk-core/src/thread/Continuation.cpp

```diff
@@ -0,0 +1,219 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include <olp/core/thread/Continuation.h>
+
+#include <atomic>
+#include <memory>
+#include <thread>
+#include <utility>
+
+#include <olp/core/client/CancellationToken.h>
+#include <olp/core/thread/SyncQueue.h>
+#include <olp/core/thread/TaskScheduler.h>
+
+namespace olp {
+namespace thread {
+
+using internal::ContinuationImpl;
+
+namespace internal {
+
+class Processor {
+  using Task = std::function<void()>;
+  struct ProcessorInternal;
+
+ public:
+  Processor(ExecutionContext execution_context,
+            std::deque<ContinuationImpl::ContinuationTask> tasks,
+            ContinuationImpl::FinalCallbackType final_callback)
+      : processor_(std::make_shared<ProcessorInternal>(
+            std::move(tasks), std::move(final_callback),
+            std::move(execution_context))) {}
+
+  // returns false when the current task is over and the queue should proceed
+  // further, true means do nothing
+  bool ProcessAsyncTask(std::shared_ptr<ProcessorInternal> process,
+                        std::pair<ContinuationImpl::AsyncTaskType,
+                                  ContinuationImpl::TaskType>& task) {
+    auto& current_task = task;
+    if (!current_task.first) {
+      // go to the next task in the queue to finilize the task
+      process->tasks_.pop();
+      return false;
+    }
+
+    auto callback = [=](void* arg) mutable {
+      // limit lifetime of the context shared_ptr to the function scope, for
+      // extra safety in case user's async function stores callback somewhere
+      // for forever.
+      auto context = process;
+      // save the arg (the identity function)
+      context->last_output_ = current_task.second(arg);
+
+      // re-enter ProcessAsyncTask() in order to finilize the task.
+      if (context->last_output_) {
+        ProcessNextTask(context);
+      }
+    };
+
+    auto func = std::move(current_task.first);
+    current_task.first = nullptr;
+    func(process->LastOutput(), std::move(callback));
+
+    return true;
+  }
+
+  void FinishQueueExecution(const std::shared_ptr<ProcessorInternal>& context) {
+    if (context->final_callback_) {
+      // the final execution context, all tasks are done or the execution
+      // is cancelled
+      context->final_callback_(context->LastOutput(), context->IsCancelled());
+      while (!context->tasks_.empty()) {
+        context->tasks_.pop();
+      }
```

Comment:
Is this not the same as calling `context->tasks_.clear()`

Reply:
replaced queue with deque to be able to clear a container in a simple way.